gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp

```diff
@@ -153,18 +154,34 @@ OGRFlatGeobufLayer::OGRFlatGeobufLayer(
     if (poSpatialRef)
         m_poSRS = poSpatialRef->Clone();
 
-    CPLDebug("FlatGeobuf", "eGType: %d", (int) eGType);
-    CPLDebug("FlatGeobuf", "m_geometryType: %d", (int) m_geometryType);
-    CPLDebug("FlatGeobuf", "m_hasZ: %d", m_hasZ);
-    CPLDebug("FlatGeobuf", "m_hasM: %d", m_hasM);
-    CPLDebug("FlatGeobuf", "m_hasT: %d", m_hasT);
+    bCreateSpatialIndexAtClose = CPLFetchBool( papszOptions, "SPATIAL_INDEX", true );
+
+    CPLDebug("FlatGeobuf", "geometryType: %d, hasZ: %d, hasM: %d, hasT: %d", (int) m_geometryType, m_hasZ, m_hasM, m_hasT);
 
     SetMetadataItem(OLMD_FID64, "YES");
 
     m_poFeatureDefn = new OGRFeatureDefn(pszLayerName);
     SetDescription(m_poFeatureDefn->GetName());
     m_poFeatureDefn->SetGeomType(eGType);
     m_poFeatureDefn->Reference();
+
+    if (bCreateSpatialIndexAtClose) {
+        CPLDebug("FlatGeobuf", "Spatial index requested will write to temp file and do second pass on close");
+        const CPLString osDirname(CPLGetPath(m_osFilename.c_str()));
+        const CPLString osBasename(CPLGetBasename(m_osFilename.c_str()));
+        m_oTempFile = CPLFormFilename(osDirname, osBasename, nullptr);
+        m_oTempFile += "_temp.fgb";
+        m_poFpWrite = VSIFOpenL(m_oTempFile.c_str(), "wb");
+    } else {
+        CPLDebug("FlatGeobuf", "No spatial index will write directly to output");
+        m_poFpWrite = VSIFOpenL(m_osFilename.c_str(), "wb");
+    }
+    if (m_poFpWrite == nullptr) {
```

Comment:
ICreateLayer() should also be modified to fail when m_poFpWriter == nullptr, otherwise nullptr dereference will happen in CreateFeature() etc.   Perhaps move that creating of the file in a bool OpenTemporaryFile() method ?

Reply:
which probably calls to extend the autotest to include an attempt to create "/i_do/not_exist.fgb"

Reply:
Fixed, asssuming you meant ICreateFeature.

Reply:
I'd prefer if ICreateLayer() failed right away when m_poFpWriter == nullptr instead of differing that check to ICreateFeature()

Reply:
Makes sense now, thanks.